src/server/lock_manager/waiter_manager.rs

```diff
@@ -242,69 +439,71 @@ impl WaiterManager {
-                .then(move |_| {
-                    wait_table
-                        .borrow_mut()
-                        .remove_waiter(start_ts, lock)
-                        .and_then(|waiter| {
-                            // The corresponding `WaitForEntry` in deadlock detector
-                            // will be removed by expiration.
-                            waiter.cb.execute(waiter.pr);
-                            Some(())
-                        });
-                    Ok(())
+    fn handle_wait_for(&mut self, handle: &Handle, waiter: Waiter) {
+        let (waiter_ts, lock) = (waiter.start_ts, waiter.lock);
+        let wait_table = self.wait_table.clone();
+        let detector_scheduler = self.detector_scheduler.clone();
+        // Remove the waiter from wait table when it times out.
+        let f = waiter.on_timeout(move || {
+            wait_table
+                .borrow_mut()
+                .remove_waiter(lock, waiter_ts)
+                .and_then(|waiter| {
+                    detector_scheduler.clean_up_wait_for(waiter.start_ts, waiter.lock);
+                    waiter.notify();
+                    Some(())
                 });
-            handle.spawn(timer);
-        }
+        });
+        self.wait_table
+            .borrow_mut()
+            .add_waiter(waiter)
+            .and_then(|old| {
+                old.notify();
+                Some(())
+            });
+        handle.spawn(f);
     }
 
-    fn handle_wake_up(
-        &mut self,
-        handle: &Handle,
-        lock_ts: TimeStamp,
-        hashes: Vec<u64>,
-        commit_ts: TimeStamp,
-    ) {
-        let mut ready_waiters = self
-            .wait_table
-            .borrow_mut()
-            .take_ready_waiters(lock_ts, hashes);
-        ready_waiters.sort_unstable_by_key(|waiter| waiter.start_ts);
-
-        for (i, waiter) in ready_waiters.into_iter().enumerate() {
-            self.detector_scheduler
-                .clean_up_wait_for(waiter.start_ts, waiter.lock);
-            if self.wake_up_delay_duration > 0 {
-                // Sleep a little so the transaction with small start_ts will more likely get the lock.
-                let when = Instant::now()
-                    + Duration::from_millis(self.wake_up_delay_duration * (i as u64));
-                let timer = Delay::new(when)
-                    .and_then(move |_| {
-                        wake_up_waiter(waiter, commit_ts);
-                        Ok(())
-                    })
-                    .map_err(|e| info!("wake-up timer delay errored"; "err" => ?e));
-                handle.spawn(timer);
-            } else {
-                wake_up_waiter(waiter, commit_ts);
+    fn handle_wake_up(&mut self, lock_ts: TimeStamp, hashes: Vec<u64>, commit_ts: TimeStamp) {
+        let mut wait_table = self.wait_table.borrow_mut();
+        let (mut visited, initial_len) = (0, wait_table.len());
```

Comment:
Return if wait_table is empty?

Reply:
It's seldom empty because only wake up when waiter count is not zero. Nevertheless I can add the check here.